operator/cmd/mesh/manifest_shared_test.go

```diff
@@ -28,6 +62,233 @@ const (
+	log.Infof("Recreating kubebuilder test environment\n")
+
+	if testenv != nil {
+		testenv.Stop()
+	}
+
+	var err error
+	testenv = &envtest.Environment{}
+	testRestConfig, err = testenv.Start()
+	if err != nil {
+		return err
+	}
+
+	testK8Interface, err = kubernetes.NewForConfig(testRestConfig)
+	testRestConfig.QPS = 50
+	testRestConfig.Burst = 100
+	if err != nil {
+		return err
+	}
+
+	s := scheme.Scheme
+	s.AddKnownTypes(v1alpha1.SchemeGroupVersion, &v1alpha1.IstioOperator{})
+
+	testClient, err = client.New(testRestConfig, client.Options{Scheme: s})
+	if err != nil {
+		return err
+	}
+
+	testReconcileOperator = istiocontrolplane.NewReconcileIstioOperator(testClient, testRestConfig, s)
+
+	return nil
+}
+
+// runManifestCommands runs all testedManifestCmds commands with the given input IOP file, flags and chartSource.
+// It returns an ObjectSet for each cmd type.
+// nolint: unparam
+func runManifestCommands(inFile, flags string, chartSource chartSourceType) (map[cmdType]*ObjectSet, error) {
+	out := make(map[cmdType]*ObjectSet)
+	for _, cmd := range testedManifestCmds {
+		log.Infof("\nRunning test command using %s\n", cmd)
+		switch cmd {
+		case cmdApply, cmdController:
+			if err := cleanTestCluster(); err != nil {
+				return nil, err
+			}
+			if err := fakeApplyExtraResources(inFile); err != nil {
+				return nil, err
+			}
+		default:
+		}
+
+		var objs *ObjectSet
+		var err error
+		switch cmd {
+		case cmdGenerate:
+			m, _, err := generateManifest(inFile, flags, chartSource)
+			if err != nil {
+				return nil, err
+			}
+			objs, err = parseObjectSetFromManifest(m)
+			if err != nil {
+				return nil, err
+			}
+		case cmdApply:
+			objs, err = fakeApplyManifest(inFile, flags, chartSource)
+		case cmdController:
+			objs, err = fakeControllerReconcile(inFile, chartSource)
+		default:
+		}
+		if err != nil {
+			return nil, err
+		}
+		out[cmd] = objs
+	}
+
+	return out, nil
+}
+
+// fakeApplyManifest runs manifest apply. It is assumed that
+func fakeApplyManifest(inFile, flags string, chartSource chartSourceType) (*ObjectSet, error) {
+	inPath := filepath.Join(testDataDir, "input", inFile+".yaml")
+	manifest, err := runManifestCommand("apply", []string{inPath}, flags, chartSource)
+	if err != nil {
+		return nil, fmt.Errorf("error %s: %s", err, manifest)
+	}
+	return NewObjectSet(getAllIstioObjects()), nil
+}
+
+// fakeApplyExtraResources applies any extra resources for the given test name.
+func fakeApplyExtraResources(inFile string) error {
+	reconciler, err := helmreconciler.NewHelmReconciler(testClient, testRestConfig, nil, nil)
+	if err != nil {
+		return err
+	}
+
+	if rs, err := readFile(filepath.Join(testDataDir, "input-extra-resources", inFile+".yaml")); err == nil {
+		if err := applyWithReconciler(reconciler, rs); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func fakeControllerReconcile(inFile string, chartSource chartSourceType) (*ObjectSet, error) {
```

Comment:
Why do we need this? It seems like we are so far off the path of the real code I am skeptical this is giving us as much coverage as we expect. Basically the entire flow of the controller is recreated here?

Reply:
The same comment applies to most of the PR, just happened to comment here. I would have expected a single place in the code switches the real client with a fake client

Reply:
The operator has three parts:
1. register/setup -> must be done with a functional API server
2. handle deletion/finalizer -> ditto
3. generate and write resources to server based on IOP -> tested by calling Reconcile() with an IOP input
The fake setup basically just skips over 1 and 2 and passes an IOP input into 3. At the end of the day it ends up exercising much of the same code as istioctl install - the difference is that path is run through the cobra command. 
I think the only reason that it's worthwhile keeping the controller test separate is it allows us to test race conditions (we can't do that with the install path since that execs a separate process). I've already found one such case which is fixed in this PR, so on the whole I think it's worth exercising this path.